binary_search_tree/BinarySearchTree.py

```diff
@@ -0,0 +1,334 @@
+            if self.left:
+                return self.left.find(value)
+            else:
+                print("Value: {} NOT FOUND!".format(value))
+                return False
+        else:
+            if self.right:
+                return self.right.find(value)
+            else:
+                print("Value: {} NOT FOUND!".format(value))
+                return False
+
+    def preorder(self):
+        if self:
+            print(str(self.data))
+            if self.left:
+                self.left.preorder()
+            if self.right:
+                self.right.preorder()
+
+    def postorder(self):
+        if self:
+            if self.right:
+                self.right.postorder()
+            if self.left:
+                self.left.postorder()
+            print(str(self.data))
+
+    def inorder(self):
+        if self:
+            if self.left:
+                self.left.inorder()
+            print(str(self.data))
+            if self.right:
+                self.right.inorder()
+
+    def __str__(self):
+        if self.data is not None:
+            return "Data: " + str(self.data)
+        return "No data in this node."
+
+
+class BST():
+    def __init__(self):
+        self.root = None
+
+    def insert(self, value):
+        if self.root:
+            return self.root.insert(value)
+        else:
+            self.root = Node(value)
+            return True
+
+    # the sole purpose of this method is to move the right sub-tree up
+    # however it does not need access to an instance "self" of class
+    # but it only makes sense if it is a part of the BST class
+    # hence its static in nature
+    @staticmethod
+    def moveRightTree(parentNode, nodeToDelete):
+        # it can't have left child so move the right sub-tree
+        if nodeToDelete.right:
+            if parentNode.data > nodeToDelete.data:
+                parentNode.left = nodeToDelete.right
+            elif parentNode.data < nodeToDelete.data:
+                parentNode.right = nodeToDelete.right
+        # it has no children
+        else:
+            if nodeToDelete.data < parentNode.data:
+                parentNode.left = None
+            else:
+                parentNode.right = None
+        del nodeToDelete
+
+    def delete(self, value):
+        # if tree is empty
+        if self.root is None:
+            return False
+
+        # data is in root node
+        elif self.root.data == value:
+            # if there are no children
+            if self.root.left is None and self.root.right is None:
+                self.root = None
+            # if only left child exists
+            elif self.root.left and self.root.right is None:
+                self.root = self.root.left
+            # if only right child exists
+            elif self.root.left is None and self.root.right:
+                self.root = self.root.right
+            # if both children exist
+            elif self.root.left and self.root.right:
```

Comment:
Can this just be an `else', as you already have checked all the possible cases earlier?

Reply:
It can be an `else` but this is repo for all algos/DS and mostly new people would clone it in order to learn python or other language. So it's best to actually explain what the `else` case is.